[lldb/Plugins] Introduce Scripted Platform Plugin

medismailben opened the pull request:
This patch introduces Scripted Platform, a new platform plugin that can be customized with a python script.

For now this can list processes described in the python script file but eventually, it will be used to spawn scripted processes and act as an interface between them.

This patch is also a less intrusive implementation of 2d53527.

It introduces a new PlatformMetadata held by every platform that contains various objects that might be used by a platform instance.

In its current form, the PlatformMetadata holds a reference to the Debugger and a ScriptedMetadata pointer. These are necessary in other to instanciate the scripted object that the ScriptedPlatform interacts with.

In order to make it less introsive with the rest of lldb's platform creation code, platform metadata are set after the platform creation, and requires to platform to reload them (using `Platform::ReloadMetadata`).

This approach has the tradeoff that the ScriptedPlaform instance is technically invalid and useless right after its creation. However, the user should never be in that situation, since we reload the platform metadata everytime with create or select the platform.

This work was previously reviewed in:
- https://reviews.llvm.org/D139252
- https://reviews.llvm.org/D142266

bulbazord commented on lldb/source/Plugins/Platform/scripted/ScriptedPlatform.cpp:
```diff
@@ -0,0 +1,288 @@
+  Debugger &debugger = m_metadata->GetDebugger();
+
+  if (!IsScriptLanguageSupported(debugger.GetScriptLanguage()))
+    return error_with_message("Debugger language not supported");
```
nit: The error message isn't really accurate right? It's not a `Debugger` language as much as it's a `Scripting` language here.

bulbazord commented on lldb/source/Plugins/Platform/scripted/ScriptedPlatform.cpp:
```diff
@@ -0,0 +1,288 @@
+
+ScriptedPlatform::ScriptedPlatform() : Platform(false) {}
+
+bool ScriptedPlatform::SetupScriptedObject(Status &error) {
```
2 Questions about this method:
1. What is a `ScriptedObject` in this context?
2. Why does this return a `bool` and take a mutable `Status &` as argument? Why doesn't it just return a `Status` object directly?

medismailben replied:
1. A Scripted Object is the class instance that we will use to make the calls to.
2. This could just return an `llvm::Error` actually.

bulbazord commented on lldb/source/Plugins/Platform/scripted/ScriptedPlatform.cpp:
```diff
@@ -0,0 +1,288 @@
+
+ScriptedPlatform::~ScriptedPlatform() {}
+
+bool ScriptedPlatform::ReloadMetadata() {
```
Why does this return a `bool`? If you want to know what happened, does it not make sense to return a `Status` object directly? Or an `llvm::Error` if that's your thing.

bulbazord commented on lldb/source/Plugins/Platform/scripted/ScriptedPlatform.cpp:
```diff
@@ -0,0 +1,288 @@
+}
+
+void ScriptedPlatform::Initialize() {
+  if (g_initialize_count++ == 0) {
```
Why does this get initialized multiple times? Once per script?

medismailben replied:
I might be wrong but IIUC, we make a new platform plugin for every new instance of a platform. Like you could have 2 different `platform remote-ios` talking with 2 different devices. That's why most platform plugins have this `g_initialize_count` checks in `Initialize & Uninitialize, to be able to make new instances of the plugins every time we connect to a new platform, but without having to register it over and over.

labath replied:
In most cases that is completely unnecessary cargo-culted code. `Initialize` is called just once, from `SBDebugger::Initialize()`. The only case where this might be useful is when a platform class is (also) used as a base class of some other platform (and the other platform delegates the Initialize call to the base). However:
- that's not the case here
- even for those cases, I think that pattern could be replaced by something cleaner

bulbazord commented on lldb/source/Plugins/Platform/scripted/ScriptedPlatform.cpp:
```diff
@@ -0,0 +1,288 @@
+ScriptedPlatform::ParseProcessInfo(StructuredData::Dictionary &dict,
+                                   lldb::pid_t pid) const {
+  if (!dict.HasKey("name"))
+    return llvm::make_error<llvm::StringError>(
```
`createStringError`?

medismailben replied:
Didn't existed back then :p, good catch!

bulbazord commented on lldb/source/Plugins/Platform/scripted/ScriptedPlatform.cpp:
```diff
@@ -0,0 +1,288 @@
+                                   lldb::pid_t pid) const {
+  if (!dict.HasKey("name"))
+    return llvm::make_error<llvm::StringError>(
+        "No 'arch' key in process info dictionary.",
```
Error message is has wrong key.

medismailben commented on lldb/source/API/SBPlatform.cpp:
```diff
@@ -299,6 +301,35 @@ SBPlatform::SBPlatform(const char *platform_name) {
+    : SBPlatform(platform_name) {
+  LLDB_INSTRUMENT_VA(this, platform_name, debugger, script_name, dict);
+
+  if (!m_opaque_sp)
```
new user-defined standalone scripted platform instances, yes

medismailben commented on lldb/source/API/SBPlatform.cpp:
```diff
@@ -299,6 +301,35 @@ SBPlatform::SBPlatform(const char *platform_name) {
+  if (!m_opaque_sp)
+    return;
+
+  if (!script_name || !dict.IsValid() || !dict.m_impl_up)
```
Actually, `SBStructuredData::IsValid` just calls `IsValid` on the `unique_ptr` without checking it before, which could lead to a crash. I'll fix that in a separate patch.

bulbazord replied:
The constructor for `SBStructuredData` creates a new empty `StructuredDataUP` in its constructor every time. That should be okay.

labath commented on lldb/test/API/functionalities/scripted_platform/TestScriptedPlatform.py:
```diff
@@ -0,0 +1,108 @@
+        """Test that we can load and select an lldb scripted platform using the
+        SBAPI, check its process ID, parent, name & triple.
+        """
+        os.environ["SKIP_SCRIPTED_PLATFORM_SELECT"] = "1"
+
+        def cleanup():
+            del os.environ["SKIP_SCRIPTED_PLATFORM_SELECT"]
+
+        self.addTearDownHook(cleanup)
```
Is it important that the HandleCommand thingy happens in __lldb_init_module. Could we remove this environment business and move the call to the relevant test instead?

labath replied:
I don't want to jump in the middle and have you change the way you do things, but I am somewhat allergic to using environment variables for passing information around, even if its just for test. Personally, I'd probably iterate on this by putting everything on an lldb command line (like, `lldb -o "command script import ..." -o "platform select scripted-platform ..."`) and then running that over and over.

Up to you, I guess...

medismailben replied:
I understand your point of view, I actually got comments from other people (@JDevlieghere) about this and I've been doing it in other scripted extension tests (i.e. `scripted process`).

Now that I'm not iterating on them anymore, I can remove these environment variables from all the tests but I prefer doing in a follow-up.

labath replied:
Sounds good. (you could also do it before hand, as I think this patch is going to take a couple more iterations).

labath commented on lldb/source/API/SBStructuredData.cpp:
```diff
@@ -86,7 +86,11 @@ lldb::SBError SBStructuredData::SetFromJSON(lldb::SBStream &stream) {
   m_impl_up->SetObjectSP(json_obj);
 
-  if (!json_obj || json_obj->GetType() != eStructuredDataTypeDictionary)
+  static constexpr StructuredDataType structured_data_record_type[] = {
```
Can we put this into a separate patch (with a test case) ?

medismailben replied:
Here: #101929

labath commented on lldb/test/API/functionalities/scripted_platform/TestScriptedPlatform.py:
```diff
@@ -0,0 +1,108 @@
+class ScriptedPlatformTestCase(TestBase):
+    NO_DEBUG_INFO_TESTCASE = True
+
+    @skipUnlessDarwin
```
What's the reason this doesn't work elsewhere?

labath commented on lldb/include/lldb/Target/Platform.h:
```diff
@@ -1003,6 +1010,21 @@ class Platform : public PluginInterface {
+  }
+
+protected:
+  Debugger &m_debugger;
```
I'm somewhat worried about how this makes a platform debugger-specific through a back door. I take it this is needed to access the script interpreter (?) What happens if this platform is later added to a different debugger object?

labath replied:
What I mean is that currently, when one creates a "normal" platform instance, it is not tied to any debugger. Later, one can add it to the platform list of any debugger (even multiple ones). This scripted platform is tied to a debugger (using this field), but it can still be inserted into the platform list of any debugger -- and it's not clear to me whether that's expected to work. In principle, I guess it can work, because there's a difference between the debugger whom you are serving, and the debugger which holds your script interpreter (and the two need not be the same in theory), but is that what you want to support?

If yes, then yeah,  I think it'd be better to hold a scriptinterpreter here.

medismailben replied:
I don't see how passing the debugger vs. the script interpreter would make a difference if the scripted platform was created with one debugger and copied to another debugger platform's list. Even if I held on the script interpreter from the first debugger, once copied to the other one, the scripted platform will still interact with the script interpreter from the first one, which is basically the same as holding to the debugger.

Am I understanding that correctly?

I'm not sure if this is a scenario we want to support, may be @jimingham would have some opinions about this.

@labath if you have some ideas how to support this scenario, I'd love to be convinced :)

labath replied:
> I don't see how passing the debugger vs. the script interpreter would make a difference if the scripted platform was created with one debugger and copied to another debugger platform's list. Even if I held on the script interpreter from the first debugger, once copied to the other one, the scripted platform will still interact with the script interpreter from the first one, which is basically the same as holding to the debugger.

It's basically the same, but it makes the intent clearer, I think.

> I'm not sure if this is a scenario we want to support, may be @jimingham would have some opinions about this.

> @labath if you have some ideas how to support this scenario, I'd love to be convinced :)

I am not saying we should support this scenario (I'm also not saying we should not). What I'm saying is that this (creating a platform with one debugger, but then adding it to another one) is something that can happen, so we should figure out what to do about it. Do we want to support it or not? Do we want to prevent it from happening or not?

I don't really have an answer to those, but I think we should have one..

jimingham replied:
If there are platforms tied to a specific debugger, we should make sure they aren't look-upable by any other Debuggers.  Maybe we need a Debugger specific list of Platforms where we put the scripted ones, and change the platform lookup so it consults the Debugger specific ones first, then the global list.  That would have the added benefit (if you do the lookup right) of being able to wrap a known platform in a scripted one to slightly modify its behavior.

medismailben replied:
Also discussed about that offline with @JDevlieghere and he suggested may be the scripted platforms should instantiate their own script interpreter so they're never dependent on a debugger. The issue with that is that now, the processes that will be create from the scripted platform will be restricted to use that script interpreter instead of using the target's debugger script interpreter. I'm not sure yet if this is desirable or even feasible but it's another lead we could take.

jimingham replied:
I don't like that idea, a scripted platform is really really going to want to create scripted processes.  That will be super awkward when the scripted platform and the scripted process can't share code because they are in different interpreters.  I don't want to have to explain this to people trying to use the lldb extension points.

I think a much simpler solution is that Debuggers have a list of Platforms which is where dynamically created platforms are registered, and which it consults first when resolving "platform for name".   Debuggers really should be independent of one another or things get hard to reason about.  For instance, it would be very strange to have `command script import` in debugger A make a new Platform show up in debugger B.   

So you want to keep platforms made by a debugger confined to that debugger, and having an "dynamic platforms" or whatever list in the debugger seems a pretty straightforward way of doing that.

labath commented on lldb/source/Commands/CommandObjectPlatform.cpp:
```diff
@@ -180,7 +184,19 @@ class CommandObjectPlatformSelect : public CommandObjectParsed {
         if (platform_sp) {
           GetDebugger().GetPlatformList().SetSelectedPlatform(platform_sp);
-
+          OptionGroupPythonClassWithDict &script_class_opts =
```
What happens if one tries to set metadata on a platform that doesn't support/care about it?

labath replied:
> Nothing.

Okay. Is that what we want to happen? Should we maybe make that an error, or at least give the user a warning that the arguments it laboriously typed are going to be ignored?

medismailben replied:
So `Platform::SetMetadata` returns void, but `Platform::ReloadMetadata` returns an `llvm::Error` as you can see down below (line 196). When setting it from the command line, the user will see an error message and the command interpreter however from the SBAPI, since this is called from the `SBPlatform` constructor, there is no way to do error reporting (without adding an overload for the constructor), so the error is consumed and logged to the `lldb platform` channel.

labath replied:
Okay, this is better, though a more user-focused error message would be even better ("platform XXX does not support scripting metadata" ?) I'm less worried about what happens in the SB API version, though if the operation can fail for more than one reason, then it might be a good idea to add an `SBError&` to the appropriate constructor.

clayborg replied:
Or don't specify the metadata on the constructor, but with an accessor that returns a SBError and omit it from the constuctor.

medismailben replied:
@clayborg I considered that but that would mean that the platform is in an invalid state after construction and it would require the user to call that accessor in order to "complete" its construction. I think adding an overload with an extra `SBError&`  would be fine here.

labath commented on lldb/source/Plugins/Platform/scripted/ScriptedPlatform.cpp:
```diff
@@ -0,0 +1,305 @@
+}
+
+std::vector<ArchSpec>
+ScriptedPlatform::GetSupportedArchitectures(const ArchSpec &process_host_arch) {
```
This doesn't seem right. Normally a platform should "know" which architectures it supports without having access to any specific process. In this case, that would probably mean calling out to python so that the plugin author can specify those. `process_host_arch` was only added to disambiguate some cases (and I don't remember what those cases were, but I think it had something to do with macos simulator platforms). This may be fine for now, but we probably shouldn't rely on this implementation...

labath replied:
Yes, I think that's basically it.

medismailben committed:
[lldb/API] Fix SBStructuredData JSON Array parsing

This patch loosen the parsing requirement to allow parsing not only
JSON dictionaries but also JSON arrays.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

clayborg commented on lldb/bindings/interface/SBPlatformDocstrings.i:
```diff
@@ -29,3 +29,15 @@ and executable type. If the architecture or executable type do not match,
+:param platform_name: name of the platform plugin to use to create the platform
+:param debugger: debugger instance owning the platform
+:param script_name: name of the script class and module to use to create the platform
+:param dict: user-provided dictionary that can be used when instanciating a platform
```
```
s/instanciating/instantiating/
```

clayborg commented on lldb/include/lldb/API/SBPlatform.h:
```diff
@@ -100,6 +99,9 @@ class LLDB_API SBPlatform {
 
   SBPlatform(const char *platform_name);
 
+  SBPlatform(const char *platform_name, const SBDebugger &debugger,
+             const char *script_name, const SBStructuredData &dict);
+
```
is `script_name` the name of a class or a function? If it is a class that implements the platform API, then should we avoid passing in `platform_name` and have that name be discovered by a call to the platform class itself?

medismailben replied:
That sounds like a good idea, I'll add a getter to the python base class.

clayborg commented on lldb/include/lldb/API/SBStructuredData.h:
```diff
@@ -111,6 +111,7 @@ class SBStructuredData {
 protected:
   friend class SBAttachInfo;
   friend class SBLaunchInfo;
+  friend class SBPlatform;
   friend class SBDebugger;
   friend class SBTarget;
   friend class SBProcess;
```
We keep friending new classes in all of the `lldb::SB` header files. I wonder if we should create a single class inside of lldb that has all of the abilities to access the private methods of any `lldb::SB` classes and then have a bunch of static calls on this one class that everyone internally can use?

medismailben replied:
That would be nice, but I think we should do this in a different patch. I'll keep in mind though 🙂

clayborg commented on lldb/test/API/functionalities/scripted_platform/TestScriptedPlatform.py:
```diff
@@ -0,0 +1,108 @@
+        structured_data.SetFromJSON(json.dumps({"processes": [proc_info]}))
+
+        platform = lldb.SBPlatform(
+            "scripted-platform",
```
Repeating what I said above, but can we remove this and have the `my_scripted_platform.MyScriptedPlatform` have a `def get_platform_name(self):` method that must be implemented? This way we won't ever have a plugin initialize itself with a name that it won't ever know about? Or the `my_scripted_platform.MyScriptedPlatform.__init__` method could get this name passed down to it so it can hold onto it?

medismailben replied:
Replied in https://github.com/llvm/llvm-project/pull/99814#discussion_r1706399944

clayborg commented on lldb/test/API/functionalities/scripted_platform/my_scripted_platform.py:
```diff
@@ -6,20 +6,27 @@
+            for i in range(0, processes.GetSize()):
+                proc_info = processes.GetItemAtIndex(i)
+                proc = {}
+                proc["name"] = proc_info.GetValueForKey("name").GetStringValue(42)
+                proc["arch"] = proc_info.GetValueForKey("arch").GetStringValue(42)
+                proc["pid"] = proc_info.GetValueForKey("pid").GetIntegerValue()
+                proc["parent"] = proc_info.GetValueForKey("parent").GetIntegerValue()
+                proc["uid"] = proc_info.GetValueForKey("uid").GetIntegerValue()
+                proc["gid"] = proc_info.GetValueForKey("gid").GetIntegerValue()
```
Can we document this dictionary format somewhere that users can see this and specify which ones are mandatory and which are optional?

medismailben replied:
This is document on the `scripted_platform.py` base class and once this lands, the website will automatically be updated so users can see it. [This is what it should look like.](https://lldb.llvm.org/python_api/lldb.plugins.scripted_process.ScriptedProcess.html#lldb.plugins.scripted_process.ScriptedProcess.get_loaded_images)

JDevlieghere commented:
I haven't had a chance to review this, but I would appreciate if we can keep in mind how this might affect breaking up the platforms. Right now, the Platform abstraction covers two concepts: 

1. How you connect to a platform and do things like transfer files, list processes, etc.
2. How to debug on a certain platform, which includes things like SDK, shared cache, etc

We ran into this problem when adding the (downstream) `device` command. We had new way to talk to a device (using a private framework) and I really didn't want to create a new macOS, iOS, watchOS, etc platform just for that. So for Xcode 16 we worked around it, but I really want to go back and do it the right way in the near future. I think it should be possible to detangle these two concepts so I want to make sure this patch doesn't make that harder.

clayborg commented:
Is part of the goal of this class to extend another platform by being able to override a built in platform? If so, we might want the script platform class to be able to specify the name of the platform it wants to extend with a method. Something like:
```
class ScriptedPlatform:
  def extends_platform_name(self):
    '''Get the name of the internal platform this class extends.'''
    return "iPhoneOS.platform"
  def get_platform_name(self):
    '''Get the name of this platform.'''
    return "iPhoneOS.platform.scripted"
```
Then the script platform python class would instantiate a `"iPhoneOS.platform"` platform and keep it as an instance variable, then the ScriptedPlatformPython would check if the python class implements the method in python, and if it doesn't, it will call through to the `"iPhoneOS.platform"` platform via its instnace variable. Any methods that are overridden in the scripted pyton class would take precedence and not call through to the underlying platform. If a python class doesn't have a `def extends_platform_name(self):` method, then any methods that aren't there just return an error of `"unimplemented"`.

clayborg commented:
The calling through to the extended platform would happen in C++ land, not in python...

medismailben commented:
> Is part of the goal of this class to extend another platform by being able to override a built in platform? If so, we might want the script platform class to be able to specify the name of the platform it wants to extend with a method. Something like:
> 
> ```
> class ScriptedPlatform:
>   def extends_platform_name(self):
>     '''Get the name of the internal platform this class extends.'''
>     return "iPhoneOS.platform"
>   def get_platform_name(self):
>     '''Get the name of this platform.'''
>     return "iPhoneOS.platform.scripted"
> ```
> 
> Then the script platform python class would instantiate a `"iPhoneOS.platform"` platform and keep it as an instance variable, then the ScriptedPlatformPython would check if the python class implements the method in python, and if it doesn't, it will call through to the `"iPhoneOS.platform"` platform via its instnace variable. Any methods that are overridden in the scripted pyton class would take precedence and not call through to the underlying platform. If a python class doesn't have a `def extends_platform_name(self):` method, then any methods that aren't there just return an error of `"unimplemented"`.

Very interesting idea. This is not was I was thinking of while implementing this: `ScriptedPlatform` is meant mostly to mock a new platform to be able to list and "launch" or "attach" to processes (mainly `ScriptedProcess` but nothing should prevent us to launch a `GDBRemote` process or any other kind.

I think this is a pretty cool idea, but this PR is already pretty big & intrusive, so I think we should do it as a follow-up, once we've figured out how we could break up platforms like @JDevlieghere explained.

I'd also love to hear other people opinion on extending platforms through ScriptedPlatform (cc @labath @jimingham @JDevlieghere )

medismailben committed:
[lldb/Plugins] Introduce Scripted Platform Plugin

This patch introduces Scripted Platform, a new platform plugin that can
be customized with a python script.

For now this can list processes described in the python script file but
eventually, it will be used to spawn scripted processes and act as an
interface between them.

This patch is also a less intrusive implementation of 2d53527.

It introduces a new PlatformMetadata held by every platform that
contains various objects that might be used by a platform instance.

In its current form, the PlatformMetadata holds a reference to the
Debugger and a ScriptedMetadata pointer. These are necessary in other to
instanciate the scripted object that the ScriptedPlatform interacts with.

In order to make it less introsive with the rest of lldb's platform
creation code, platform metadata are set after the platform creation,
and requires to platform to reload them (using `Platform::ReloadMetadata`).

This approach has the tradeoff that the ScriptedPlaform instance is
technically invalid and useless right after its creation. However, the user
should never be in that situation, since we reload the platform metadata
everytime with create or select the platform.

This work was previously reviewed in:
- https://reviews.llvm.org/D139252
- https://reviews.llvm.org/D142266

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

labath commented:
> I'd also love to hear other people opinion on extending platforms through ScriptedPlatform (cc @labath @jimingham @JDevlieghere )

I think that having the ability to forward operations to existing platform instances would be very useful. If you look at PlatformQemuUser, most of its functions just forward operations to be host platform. We also have another downstream plugin which looks very similar to that. I would love if both of these could be lightweight python wrappers.

That said, I'm not exactly sure how Greg's idea would work, or if it's even needed. Some of this functionality without any special support, by just having an `SBPlatform` member inside the python class. This would be more flexible as you could call it only when it suits you (going with the Qemu example, the emulator mostly exposes the host filesystem, but it has some special remapping logic for system libraries (like LD_LIBRARY_PATH, but implemented in the emulator), so we may want to implement get/putfile mostly by deferring to the host, and only doing something custom for special paths). And if you really wanted to you could make a python wrapper class that makes all of this looks like (python) inheritance -- without any special support in lldb.

The main flaw in this idea is that the SBPlatform interface is pretty small compared to the internal Platform interface. I'm guessing the python class will be implementing something like the internal interface, while it will only have the external API to work with...

jimingham commented:
> > I'd also love to hear other people opinion on extending platforms through ScriptedPlatform (cc @labath @jimingham @JDevlieghere )
> 
> I think that having the ability to forward operations to existing platform instances would be very useful. If you look at PlatformQemuUser, most of its functions just forward operations to be host platform. We also have another downstream plugin which looks very similar to that. I would love if both of these could be lightweight python wrappers.
> 
> That said, I'm not exactly sure how Greg's idea would work, or if it's even needed. Some of this functionality without any special support, by just having an `SBPlatform` member inside the python class. This would be more flexible as you could call it only when it suits you (going with the Qemu example, the emulator mostly exposes the host filesystem, but it has some special remapping logic for system libraries (like LD_LIBRARY_PATH, but implemented in the emulator), so we may want to implement get/putfile mostly by deferring to the host, and only doing something custom for special paths). And if you really wanted to you could make a python wrapper class that makes all of this looks like (python) inheritance -- without any special support in lldb.
> 
> The main flaw in this idea is that the SBPlatform interface is pretty small compared to the internal Platform interface. I'm guessing the python class will be implementing something like the internal interface, while it will only have the external API to work with...

For Greg's idea to be useful, you mostly don't want to override the platform you are extending, you want to do some stuff before the underlying platform call, then you want to do the platform call, then you potentially want to do some more work afterwards.  So it would have to be something more like:

```
class ScriptedPlatform:
  def extends_platform_name(self):
    '''Get the name of the internal platform this class extends.'''
    return "iPhoneOS.platform"
  def extend_before_attach_to_process(self, attach_info):
     '''Do something to prepare for the attach"
  def extend_after_attach_to_process(self, process):
     '''Do something after the platform attach was complete'''

```
And maybe:

```
  def override_list_processes():
  '''Take over the listing of processes from the underlying Platform'''

```
But particularly for the before & after wrapping, that seems like it would be awkward to read and write, since you have to partition the before & after into different methods.

We should look at what methods the scripted platform has/will have.  For the most part the ones there now look like things that it would be quite reasonable to ask the SBPlatform to do through the API.  If that is true, then I think allowing the scripted platform to make an instance of the SBPlatform it wants to extend and just dispatch to that directly in Python would be a more straightforward design.

clayborg commented:
To extend on Pavel's modification of my idea:
```
class ScriptedPlatform:
  def extends_platform_name(self):
    '''Get the name of the internal platform this class extends.'''
    return "iPhoneOS.platform"
```
Then every function would have 3 variants:
```
def pre_XXX(...):
def XXX(...)
def post_XXX(...)
```
So for "attach_to_process" this would mean the possible functions are:
```
class ScriptedPlatform:
  def pre_attach_to_process(self, ...):
    # Called before the extended platform default attach_to_process() functionality
    # but only if `def extends_platform_name(self)` exists in this class. The default
    # attach_to_process() from the extended platform will be called after this function
    # completes.

  def attach_to_process(self, ...):
    # This will override the attach_to_process() functionality. Clients that implement
    # this function should not override pre_attach_to_process() or 
    # post_attach_to_process() as everything can be done in this function.

  def post_attach_to_process(...)
    # Called after the extended platform's default attach_to_process() functionality
    # but only if `def extends_platform_name(self)` exists in this class. 
```

jimingham commented:
> To extend on Pavel's modification of my idea:
> 
> ```
> class ScriptedPlatform:
>   def extends_platform_name(self):
>     '''Get the name of the internal platform this class extends.'''
>     return "iPhoneOS.platform"
> ```
> 
> Then every function would have 3 variants:
> 
> ```
> def pre_XXX(...):
> def XXX(...)
> def post_XXX(...)
> ```
> 
> So for "attach_to_process" this would mean the possible functions are:
> 
> ```
> class ScriptedPlatform:
>   def pre_attach_to_process(self, ...):
>     # Called before the extended platform default attach_to_process() functionality
>     # but only if `def extends_platform_name(self)` exists in this class. The default
>     # attach_to_process() from the extended platform will be called after this function
>     # completes.
> 
>   def attach_to_process(self, ...):
>     # This will override the attach_to_process() functionality. Clients that implement
>     # this function should not override pre_attach_to_process() or 
>     # post_attach_to_process() as everything can be done in this function.
> 
>   def post_attach_to_process(...)
>     # Called after the extended platform's default attach_to_process() functionality
>     # but only if `def extends_platform_name(self)` exists in this class. 
> ```

If we were to do it this way, I would not have "attach_to_process" be the override of the underlying platform's method, because you could also have a scripted platform that extends nothing, and it would use attach_to_process but override nothing.  I think it would be clearer to have the "override" version explicitly state that it is that, so: `override_attach_to_process` rather than `attach_to_process`.

But again, if most of the things that we want the scripted platform to do are methods that really ought to be available in the SBPlatform API, then I think it would be much more straightforward to be able to have the scripted version just make an instance of the platform it wants to wrap, and do the dispatch itself.

I do think it would also be handy to be able to have a wrapping scripted platform override the name lookup of the platform it wraps as well.  You might have something special you wanted to do for you installation with say the `remote-macosx` platform in your environment, and it would be simpler if you could just tell people to install your scripted platform, and then just use `remote-macosx` as they were before.  That means we need to have a shadowing list (in the Debugger) of the scripted platforms, and also a way to ask for the "built-in" platform of that name.
